lib/providers/ipmeta_provider_maxmind_v2.c

```diff
@@ -0,0 +1,914 @@
+      directory = optarg;
+      break;
+
+    case 'l':
+      state->locations_file = strdup(optarg);
+      break;
+
+    case '?':
+    case ':':
+    default:
+      usage(provider);
+      return -1;
+    }
+  }
+
+  if (directory != NULL) {
+    /* check if they were daft and specified explicit files too */
+    if (state->locations_file != NULL || state->blocks_file != NULL) {
+      fprintf(stderr, "WARNING: both directory and file name specified.\n");
+
+      /* free up the dup'd strings */
+      if (state->locations_file != NULL) {
+        free(state->locations_file);
+        state->locations_file = NULL;
+      }
+
+      if (state->blocks_file != NULL) {
+        free(state->blocks_file);
+        state->blocks_file = NULL;
+      }
+    }
+
+    /* remove the trailing slash if there is one */
+    if (directory[strlen(directory) - 1] == '/') {
+      directory[strlen(directory) - 1] = '\0';
+    }
+
+    /* malloc storage for the dir+/+file string */
+    if ((state->locations_file = malloc(
+           strlen(directory) + 1 + strlen(LOCATIONS_FILE_NAME) + 1)) == NULL) {
+      ipmeta_log(__func__, "could not malloc location file string");
+      return -1;
+    }
+
+    if ((state->blocks_file = malloc(strlen(directory) + 1 +
+                                     strlen(BLOCKS_FILE_NAME) + 1)) == NULL) {
+      ipmeta_log(__func__, "could not malloc blocks file string");
+      return -1;
+    }
+
+    /** @todo make this check for both .gz and non-.gz files */
+
+    ptr = stpncpy(state->locations_file, directory, strlen(directory));
+    *ptr++ = '/';
+    /* last copy needs a +1 to get the terminating nul. d'oh */
+    ptr = stpncpy(ptr, LOCATIONS_FILE_NAME, strlen(LOCATIONS_FILE_NAME) + 1);
+
+    ptr = stpncpy(state->blocks_file, directory, strlen(directory));
+    *ptr++ = '/';
+    ptr = stpncpy(ptr, BLOCKS_FILE_NAME, strlen(BLOCKS_FILE_NAME) + 1);
+  }
+
+  if (state->locations_file == NULL || state->blocks_file == NULL) {
+    fprintf(stderr, "ERROR: %s requires either '-d' or both '-b' and '-l'\n",
+            provider->name);
+    usage(provider);
+    return -1;
+  }
+
+  return 0;
+}
+
+/* Parse a maxmind_v2 location cell */
+static void parse_maxmind_v2_location_cell(void *s, size_t i, void *data)
+{
+  ipmeta_provider_t *provider = (ipmeta_provider_t *)data;
+  ipmeta_provider_maxmind_v2_state_t *state = STATE(provider);
+  ipmeta_record_t *tmp = &(state->tmp_record);
+  char *tok = (char *)s;
+
+  char *end;
+
+  /* skip the first line */
+  if (state->current_line < HEADER_ROW_CNT) {
+    return;
+  }
+
+
+  switch (state->current_column) {
+  case LOCATION_COL_ID:
+    /* init this record */
+    tmp->id = strtol(tok, &end, 10);
+    if (end == tok || *end != '\0' || errno == ERANGE) {
+      ipmeta_log(__func__, "Invalid ID Value (%s)", tok);
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    break;
+
+  case LOCATION_COL_LOCALCODE:
```

Comment:
please be careful with "local" vs "locale". pick the correct one and use it consistently.